### DISCUSSION TITLE
Inquiry regarding Prior Covariance Matrix construction and weighting for Posterior Covariance Matrix calculation

### DISCUSSION THREAD

**Discussion by g3rm7** [gmdsi/GMDSI_notebooks#123] [General]
HI. Firstly, I extend my appreciation for your diligent work and the provision of the notebooks, which have proven immensely beneficial.

However, I find myself encountering some confusion regarding the construction of the prior covariance matrix, the determination of the standard deviation of measurement noise for weighting purposes, and the subsequent implications for the calculation of the posterior covariance matrix. I have articulated the specific questions below:

1. Prior Covariance Matrix:
- How does one determine the proper variables (e.g., sill, nugget, range) for variograms used in the construction of the prior covariance matrix?
- Could you provide insights into the units associated with these variables and offer guidance on their integration when dealing with different types of adjustable parameters?

2. Measurement Noise:
- What constitutes "measurement noise," and what methodologies are recommended for obtaining its standard deviation within a dataset?
- It has been observed in the notebooks and other GMDSI tutorials that the standard deviation of measurement noise is utilized for weighting purposes in uncertainty analysis. What is the reason behind this?

3. Posterior Covariance Matrix:
- What are the potential implications for the composition of the posterior covariance matrix using a Jacobian built with history-matching weighting instead of a "correct weighting" based on standard deviation of measurement noise?
- Furthermore, how might the use of history-matching weighting in constructing the posterior covariance matrix affect the generation of parameter distributions for Monte Carlo analysis?

Thank you for your time and assistance in addressing my queries.

**Comment by mnfienen** [comment]
Hey there @g3rm7  - these are good and important questions. Here's my take:

On point 1. The role of the prior covariance is generally to enforce some continuity in parameters that are autocorrelated. My vague choice of words is intentional there - it's less important that there be a strictly tuned covariance matrix and more important that continuity in general is applied. This is kind of like the [empirical Bayes](https://agupubs.onlinelibrary.wiley.com/doi/pdf/10.1029/2007WR006730) approach in which a variogram is chosen but the parameters for it are actually tuned in the inversion. As a result of this, the units of the variogram parameters are not vital to consider. 

Despite what I said above, it is still not a bad idea to fit a variogram to data in the parameters if that's possible. It just doesn't need to be dogmatic. There's some more about this in this [reference](https://pubs.usgs.gov/sir/2010/5168/pdf/sir20105168.pdf) as well.

On point 2. Measurement noise is an interesting concept and there's a bit to unpack there. I'll start with your second question,. Indeed, the reciprocal of measurement standard deviation is used (initially) in assigning weights to measurements (observations). This dates back to linear regression theory (check out this [classic](https://pubs.usgs.gov/of/1985/0180/report.pdf) treatment for 
background). The idea was that the variability (as std) was a metric of "quality" or "importance" for a measurement. But even in that 1980s report there was acknowledgement (see page 122) that weights  transcend interpretation as only the reciprocal of variability but also control which measurements we modelers subjectively think should have more or less emphasis in the regression. So, it's not dogmatic that the weights are reciprocal of measurement noise, but it's a starting point. In the notebooks, you will see the subjectivity in weighting goes pretty far once we talk about reweighting the objective function by groups.

So.....what does this measurement noise really represent? It's a mix of instrument errors, model errors, intrinsic variability of the system maybe not even captured by the instrument......all those things. So really, I like to think of this noise as "how close do I judge correspondence between model outputs and measurements should be?" Too close and we overfit, degrading forecast ability. Too far and the model isn't adequately representing / simulating the system. So, it's subjective in the bottom line.

Finally, point 3. The posterior covariance is, by definition, impacted both by the prior and by what is learned through history matching. The implication is that it represents the best state of knowledge given what you knew before (prior) and what is supported by information mapped from measurements to the parameters (history matching).  That's a good thing, and if the weights for posterior Monte Carlo differ from the prior, that represents an important update from what you learned meaning the history matching was a worthwhile endeavor. From a Bayesian perspective, that balance is useful and remember the posterior is "conditioned" on both the prior and the measurements, so if you change either of those, you might also change the prior.

@jtwhite79 might have more to say, but hopefully that helps?

**Reply by jtwhite79** [reply]
I'd just add that on the variogram for the prior covariance matrix bit, I have rarely seen enough point data to do formal variogram modeling (very well).  But more importantly, you can still visualize realizations of parameters drawn from the covariance matrix to see if the spatial pattern of heterogeneity is plausible (so make maps of the property fields from a few reals and show them to everyone involved to make sure they are ok with them).  You dont have to run the model to do this but it can be important for building consensus early...

Note: if you are using pyEMU's PstFrom, the parameter bounds you supply are used to set the sill of the variogram for spatially distributed (if you pass the args to indicate you want geostatistical correlation) so that the bounds are coherent with the resulting range of values in the prior parameter realizations.

**Reply by laat0003** [reply]
"..you can still visualize realizations of parameters drawn from the covariance matrix to see if the spatial pattern of heterogeneity is plausible (so make maps of the property fields from a few reals and show them to everyone involved to make sure they are ok with them). You dont have to run the model to do this but it can be important for building consensus early..."

This is a great tip. I need to try and build this into my workflow.

**Comment by lesliegg1** [comment]
This thread cleared up some of my questions about the prior covariance. Just to be sure I'm understanding the mechanics right, suppose I use PstFrom.add_parameters() with a variogram model specified in the geostruct argument. Then the prior parameter realizations generated by pestpp-ies will adhere to the specified variogram model for that set of spatially correlated parameters through the prior covariance matrix, is that correct? My followup question, is there a way to incorporate uncertainty in the range parameter (for example), to allow for prior parameter realizations to reflect uncertainty in the covariance structure? Along those lines, have you ever had success with purposefully omitting a prior covariance structure and finding it through the history matching process?

**Reply by mnfienen** [reply]
Hi @lesliegg1 - you are correct. The prior parameter realizations will adhere to the structure you specify making the PstFrom object. 

On the second point, there is not a built-in way to make the range parameter uncertain. You could, however, incorporate the so-called "hyperparameters" into the loop of the history matching process, effectively making them into parameters (as far as PEST is concerned). But, in doing so, you'd probably only be able (practically) to make one draw per hyperparameter value. In my experience, making a large number of draws even with a single range parameter usually provides enough variabliity that can then be reduced through history matching so the added value of that extra parameterization may be limited. I have done it in the past, though (it's called "Empirical Bayes" when you jointly estimate hyperparameters and state parameters). More background here - https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2007WR006730 - but not sure I'd recommend it now :).

**Reply by jtwhite79** [reply]
There is a hack that Ive tried before on this front:  generate several smaller ensembles, each with a unique variogram and then concat them into a single bigger ensemble and then feed that to ies as the prior ensemble - I did this experiment with the anisotropy bearing/angle - half the reals had bearing = 45 and other had bearing=135.  It seemed to work, but there was some "bleed" in the posterior realizations (in that the 45s had some 135 degree characteristics and visa-versa...) - top is two prior reals and bottom are the same after a few ies iterations and the markers are places with HK data:

![image](https://github.com/user-attachments/assets/1d83da3e-93af-4fcf-a22a-8010e21a1591)

**Reply by laat0003** [reply]
Ha! That is exactly what I did (with different hyperparams) to get a feel for the significance of including vs excluding them as estimable. This whole thread is great, thanks all.
